pkg/controller/velero/velero.go

```diff
@@ -0,0 +1,233 @@
+			// Specs aren't equal, update and fix.
+			reqLogger.Info("Updating BackupStorageLocation")
+			foundBsl.Spec = *bsl.Spec.DeepCopy()
+			if err = r.client.Update(context.TODO(), foundBsl); err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+	}
+
+	// Install VolumeSnapshotLocation
+	foundVsl := &velerov1.VolumeSnapshotLocation{}
+	vsl := veleroInstall.VolumeSnapshotLocation(namespace, strings.ToLower(string(platformStatus.Type)), locationConfig)
+	if err = r.client.Get(context.TODO(), types.NamespacedName{Namespace: namespace, Name: "default"}, foundVsl); err != nil {
+		if errors.IsNotFound(err) {
+			// Didn't find VolumeSnapshotLocation
+			reqLogger.Info("Creating VolumeSnapshotLocation")
+			if err := controllerutil.SetControllerReference(instance, vsl, r.scheme); err != nil {
+				return reconcile.Result{}, err
+			}
+			if err = r.client.Create(context.TODO(), vsl); err != nil {
+				return reconcile.Result{}, err
+			}
+		} else {
+			return reconcile.Result{}, err
+		}
+	} else {
+		// VolumeSnapshotLocation exists, check if it's updated.
+		if !reflect.DeepEqual(foundVsl.Spec, vsl.Spec) {
+			// Specs aren't equal, update and fix.
+			reqLogger.Info("Updating VolumeSnapshotLocation")
+			foundVsl.Spec = *vsl.Spec.DeepCopy()
+			if err = r.client.Update(context.TODO(), foundVsl); err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+	}
+
+	// Install CredentialsRequest
+	foundCr := &minterv1.CredentialsRequest{}
+	cr := credentialsRequest(namespace, credentialsRequestName, instance.Status.S3Bucket.Name)
+	if err = r.client.Get(context.TODO(), types.NamespacedName{Namespace: namespace, Name: credentialsRequestName}, foundCr); err != nil {
+		if errors.IsNotFound(err) {
+			// Didn't find CredentialsRequest
+			reqLogger.Info("Creating CredentialsRequest")
+			if err := controllerutil.SetControllerReference(instance, cr, r.scheme); err != nil {
+				return reconcile.Result{}, err
+			}
+			if err = r.client.Create(context.TODO(), cr); err != nil {
+				return reconcile.Result{}, err
+			}
+		} else {
+			return reconcile.Result{}, err
+		}
+	} else {
+		// CredentialsRequest exists, check if it's updated.
+		if !reflect.DeepEqual(foundCr.Spec, cr.Spec) {
+			// Specs aren't equal, update and fix.
+			reqLogger.Info("Updating CredentialsRequest")
+			foundCr.Spec = *cr.Spec.DeepCopy()
+			if err = r.client.Update(context.TODO(), foundCr); err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+	}
+
+	// Install Deployment
+	foundDeployment := &appsv1beta1.Deployment{}
+	deployment := veleroInstall.Deployment(namespace, veleroInstall.WithoutCredentialsVolume(),
+		veleroInstall.WithEnvFromSecretKey(strings.ToUpper(awsCredsSecretIDKey), credentialsRequestName, awsCredsSecretIDKey),
+		veleroInstall.WithEnvFromSecretKey(strings.ToUpper(awsCredsSecretAccessKey), credentialsRequestName, awsCredsSecretAccessKey),
+		veleroInstall.WithImage("gcr.io/heptio-images/velero:v1.0.0"),
```

Comment:
Can we abstract the image string to a constant or some other thing? Perhaps in the Velero CR to request a specific version?

Reply:
Good idea